docs/design.md

```diff
@@ -0,0 +1,371 @@
+```
+
+1. This method takes a shard key and a `Mount`.
+2. It initializes the shard in `Unavailable` state.
+3. Calls `mount.Info()` to determine if the mount is of local or remote type.
+4. Calls `mount.Stat()` to determine if the mount target exists. If not, it
+   errors the registration.
+5. If remote, it fetches the remote resource into the scrap area.
+6. It determines the kind of CAR it is, and populates the index accordingly.
+7. Sets the state to `Available`.
+8. Returns.
+
+This method is _synchronous_. It returns only when the shard is fully indexed
+and available for serving data. This embodies the consistency property of the
+ACID model, whereby an INSERT is immediately queriable upon return.
+
+_RegisterOpts is an extension point._ In the future it can be used to pass in an
+unseal/decryption function to be used on access (e.g. such as when fast, random
+unseal is available).
+
+#### Shard destruction
+
+To destroy a shard, call:
+
+```go
+dagst.DestroyShard(key []byte) (destroyed bool, err error)
+```
+
+#### Other operations
+
+  * `[Pin/Unpin]Shard()`: retains the shard data in the local scrap area.
+  * `ReleaseShard()`: dispose of / release local scrap copies or other resources
+    on demand (e.g. unsealed copy).
+  * `[Lock/Unlock]Shard()`: for exclusive access.
+  * `[Incr/Decr]Shard()`: refcounting on shards.
+
+### Mounts
+
+Shards can be located anywhere, and can come and go dynamically e.g. Filecoin
+deals expire, removable media is attached/detached, or the IPFS user purges
+content.
+
+It is possible to mount shards with CARs accessible through the local
+filesystem, detachable mounts, NFS mounts, distributed filesystems like
+Ceph/GlusterFS, HTTP, FTP, etc.
+
+This versatility is provided by an abstraction called `mount.Mount`, a pluggable
+component which encapsulates operations/logic to:
+
+1. `Fetch() (io.ReadCloser, error)` Load a CAR from its origin.
+2. `Info() mount.Info` Provides info about the mount, e.g. whether it's local or
+   remote. This is used to determine whether the fetched CAR needs to be copied
+   to a scrap area.
+3. `Stat() (mount.Stat, error)` Equivalent to a filesystem stat, provides
+   information about the target of the mount: whether it exists, size, etc.
+
+When instantiating `Mount` implementations, one can provide credentials, access
+tokens, or other parameters through the implementation constructors. This is
+necessary if access to the CAR is permissioned/authenticated.
+
+**Local filesystem Mount**
+
+A local filesystem `Mount` implementation loads the CAR directly from the
+filesystem file. It is of `local` type and therefore requires no usage of scrap
+area.
+
+*This `Mount` is provided out-of-box by the DAG store.*
+
+**Lotus Mount implementation**
+
+A Lotus `Mount` implementation would be instantiated with a sector ID and a
+bytes range within the sealed sector file (i.e. the deal segment).
+
+Loading the CAR consists of calling the worker HTTP API to fetch the unsealed
```

Comment:
So, the: 

```
- Check if we already have a local/remote unsealed copy.
- Get a Reader to the local Unsealed bytes or by performing a ranged http query over the remote Unsealed bytes if it already exists.
- Unseal the local or remote sector file if we don't have an unsealed copy.
- Get a Reader to the local or remote unsealed bytes.
```

is all taken care of by the Lotus Mount implementation, right ? I am confused by the _"A transformation function may be provided in the **future** as a `RegisterOpt`that conducts the unsealing"_.  

Are you suggesting that it is the caller's responsibility to perform the Unsealing if no unsealed copy exists and have the Mount implementation simply return the Unsealed copy or can we hide that detail in the Lotus mount implementation ?

Reply:
Note: We can build out an endpoint in Markets that takes care of all the above magic and the Lotus mount implementation ONLY needs to worry about calling that endpoint.

Reply:
Yes, this is accurate 👍 

Just to note that Mounts do not manage transient copies (i.e. what you call "CAR not available locally"). I think we're both on the same page here, but wanted to confirm.